README.md
# Trexa Service

This project aims to provide a weekly Trexa 100k list, which is created from the Tranco and Alexa lists. In theory it could be larger, so file an issue if that's
useful to someone.

The canonical repo for Trexa is https://github.com/mozilla/trexa, but the merge
code has been rewritten here to be more flexible.

## Running locally

1. clone the repo
2. run `python3 -m venv env`
3. run `source env/bin/activate`
4. run `pip3 install -r requirements.txt`

And then go to the next section to run tests or start the service.

## Running the tests

`FLASK_ENV=development pytest`

## Starting the service

From the project root:

Production mode:
`FLASK_APP=trexa flask run --host=0.0.0.0`

or

Development mode:
`FLASK_ENV=development FLASK_APP=trexa flask run`

## Environment variables

The following environment variables can be defined to override defaults:

```
ZIP_DOWNLOADS_DEST
CSV_DOWNLOADS_DEST
FINAL_LIST_DEST
```

## HTTP Endpoints

This app exposes the following HTTP endpoints

* `/lists`: see all lists available for download
* `/lists/trexa-2020-05-21.csv`: download a full, single list (150,000+ sites)
* `/api/lists/2020-05-21?count=N`: download a single list, trimmed to N sites

## License

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at http://mozilla.org/MPL/2.0/

## Code of Conduct

This project and repository is governed by Mozilla's code of conduct and
etiquette guidelines. For more details please see the [Code of Conduct](./CODE_OF_CONDUCT.md) file.

CHANGELOG.md
# CHANGELOG.md

## v1.0.0 - ???
Initial commits. Implement the core featureset, and hopefully not a lot of bugs.

tests/test_api_helpers.py
import os
from trexa.api.endpoints import trim_csv

"""Tests for the API helper methods."""

test_dir = os.path.dirname(os.path.abspath(__file__))
test_csv = os.path.join(test_dir, 'fixtures',
                        'trexa-05-28-2020.csv')


def get_csv_length(desired_length):
    """Helper method for trimming."""
    f = trim_csv(test_csv, desired_length)
    new_file = []
    while True:
        try:
            new_file.append(next(f))
        except StopIteration:
            break
    return len(new_file)


def test_trimming_smaller():
    """Test that we end up with a smaller file."""
    assert get_csv_length(10) == 10
    assert get_csv_length(0) == 0


def test_trimming_none():
    """Test that passing in None returns the entire file."""
    assert get_csv_length(None) == 30
    # Test not passing in a count
    f = trim_csv(test_csv)
    new_file = []
    while True:
        try:
            new_file.append(next(f))
        except StopIteration:
            break
    assert len(new_file) == 30


def test_trimming_with_garbage():
    """Test that nothing weird happens."""
    assert get_csv_length(35) == 30
    assert get_csv_length('WHAT') == 30
    assert get_csv_length('🤬') == 30
    assert get_csv_length(hex(10)) == 30


tests/test_api_endspoints.py
import pytest

import trexa

"""Tests for the API Blueprint module."""


@pytest.fixture
def client():
    trexa.app.config.update(
        FINAL_LIST_DEST='tests/fixtures',
        TESTING=True,
    )
    with trexa.app.test_client() as client:
        yield client


def test_index(client):
    """Test that we do nothing useful for the API index."""
    rv = client.get('/api/')
    assert rv.status_code == 403


def test_api_download(client):
    """Test downloading files from the API endpoint."""
    rv = client.get('/api/lists/05-28-2020')
    assert rv.status_code == 200
    rv = client.get('/api/lists/05-28-2020?count=2')
    assert rv.status_code == 200
    rv = client.get('/api/lists/05-28-2020?lol=wat')
    assert rv.status_code == 200
    rv = client.get('/api/lists/what_in_the_world.py')
    assert rv.status_code == 404


def test_api_download_length(client):
    """Test the count param works as expected."""
    rv = client.get('/api/lists/05-28-2020?count=10')
    # pop off the last empty list item (because '\n' comes at the end)
    assert len(rv.data.split(b'\n')[:-1]) == 10
    assert b'11, login.tmall.com' not in rv.data
    rv = client.get('/api/lists/05-28-2020?a=b&count=10')
    assert len(rv.data.split(b'\n')[:-1]) == 10
    assert b'11, login.tmall.com' not in rv.data
    rv = client.get('/api/lists/05-28-2020?count=')
    assert len(rv.data.split(b'\n')[:-1]) == 30


tests/test_application_endpoints.py
import pytest

import trexa

"""Tests for the Flask application (non-API) endpoints.

Note: test_list_download relies on these tests being
invoked like so: `FLASK_ENV=development pytest`
"""


@pytest.fixture
def client():
    trexa.app.config.update(
        FINAL_LIST_DEST='tests/fixtures',
        TESTING=True,
    )
    with trexa.app.test_client() as client:
        yield client


def test_useless_homepage(client):
    """Test the index route serves the README.md"""
    rv = client.get('/')
    assert rv.status_code == 200
    assert b'Trexa Service' in rv.data


def test_lists_endpoint(client):
    """Test the lists template is served for /lists

    We point the FINAL_LIST_DEST to the fixtures file for testing."""
    rv = client.get('/lists')
    assert rv.status_code == 200
    assert b'Downloads' in rv.data
    assert b'<a download href="lists/trexa-05-28-2020.csv' in rv.data


if trexa.app.config['ENV'] == 'development':
    def test_list_download(client):
        """This should only pass for local development or tests."""
        print(client.config)
        rv = client.get('/lists/test_file_with_thirty_lines.csv')
        assert rv.status_code == 200
        assert 'text/csv' in rv.headers.get('content-type')
        assert 'attachment' in rv.headers.get('content-disposition')
        assert 'test_file_with_thirty_lines.csv' in rv.headers.get(
            'content-disposition')
        rv = client.get('/lists/garbage.pdf')
        assert rv.status_code == 404


trexa/__init__.py
import os

from flask import abort
from flask import Flask
from flask import render_template
from flask import safe_join
from flask import send_file
from flask import url_for
import markdown

from trexa.api.endpoints import api_bp

app = Flask(__name__)
app.config.from_object('trexa.config')

for blueprint in [api_bp]:
    app.register_blueprint(blueprint)


@app.route('/')
def index():
    """Create the index route.

    This module isn't intended to serve a useful website, so we just dump
    the README.md as HTML.
    """
    readme_file = open("README.md", "r")
    md_template_string = markdown.markdown(
        readme_file.read()
    )
    return md_template_string


@app.route('/lists')
def show_downloads():
    """Route to show all downloads."""
    files = os.listdir(app.config['FINAL_LIST_DEST'])
    return render_template('list.html', files=files)


if app.config['ENV'] == 'development':
    @app.route('/lists/<path:file_name>')
    def serve_downloads(file_name):
        """Dev-only route to serve downloads.

        NGINX will handle static assets on the production server."""
        file_path = safe_join(
            os.path.abspath(app.config['FINAL_LIST_DEST']), file_name)
        if not os.path.exists(file_path):
            return abort(404)
        return send_file(file_path, as_attachment=True, mimetype='text/csv')


trexa/api/endpoints.py
import os

from flask import abort
from flask import Blueprint
from flask import current_app
from flask import request
from flask import Response
from flask import safe_join

from trexa.api.helpers import trim_csv

"""Blueprint for API methods."""

api_bp = Blueprint('api_bp', __name__, url_prefix='/api',
                   template_folder='../templates')


@api_bp.route('/')
def index():
    """Nothing to see here."""
    return abort(403)


@api_bp.route('/lists/<list_date>')
def trim_file(list_date):
    """API route to serve count-limited lists.

    If there's no count arg (or its garbage), just serve the whole thing
    by sending None to trim_csv.
    """
    count = request.args.get('count')
    csv_path = safe_join(os.path.abspath(
        current_app.config['FINAL_LIST_DEST']), f'trexa-{list_date}.csv')
    if not os.path.exists(csv_path):
        return abort(404)
    return Response(trim_csv(csv_path, count=count), mimetype='text/csv')
